e2e/browser/test/e2e.playwright.ts

```diff
@@ -0,0 +1,136 @@
+  // the request would be sent before we wait on it.
+  await Promise.all([
+    // Negotiate the protocol endpoint at the gateway
+    page.waitForResponse(
+      (response) => response.url() === new URL(notificationGateway).href
+    ),
+    // Connect to the endpoint
+    page.waitForResponse(
+      (response) => response.url() !== new URL(notificationGateway).href
+    ),
+    // The websocket should be created
+    page.waitForEvent("websocket"),
+    page.click("button[data-testid=connectSocket]"),
+  ]);
+
+  // Allow React to re-render the status after state update
+  await page.waitForSelector("span[data-testid=webSocketStatus]");
+
+  await expect(
+    page.innerText("span[data-testid=webSocketStatus]")
+  ).resolves.toBe("connected");
+
+  // Disconnect
+  await Promise.all([page.click("button[data-testid=disconnectSocket]")]);
+
+  await websocket.waitForEvent("close");
+
+  await expect(
+    page.innerText("span[data-testid=webSocketStatus]")
+  ).resolves.toMatch("closed");
+});
+
+test("connecting a websocket, getting messages, and disconnecting it", async ({
+  page,
+}) => {
+  let websocket: PlayWrigtWebSocket;
+  // Navigate to the test page and log in.
+  await page.goto("/");
+  await essUserLogin(page, login, password);
+
+  // Make sure we have a reference to the websocket that gets created.
+  page.on("websocket", (ws) => {
+    websocket = ws;
+  });
+
+  // The button is only displayed when the websocket can be created.
+  await page.waitForSelector("button[data-testid=connectSocket]");
+
+  // Connect the websocket. Note that the Promis.all prevents a race condition where
+  // the request would be sent before we wait on it.
+  await Promise.all([
+    // Negotiate the protocol endpoint at the gateway
+    page.waitForResponse(
+      (response) => response.url() === new URL(notificationGateway).href
+    ),
+    // Connect to the endpoint
+    page.waitForResponse(
+      (response) => response.url() !== new URL(notificationGateway).href
+    ),
+    // The websocket should be created
+    page.waitForEvent("websocket"),
+    page.click("button[data-testid=connectSocket]"),
+  ]);
+
+  // Make sure the container can be created.
+  await page.waitForSelector("button[data-testid=createContainer]");
+  await Promise.all([
+    // Wait for the resource creation
+    page.waitForResponse((response) => response.status() === 201),
+    // The resource creation should trigger a websocket frame
+    websocket.waitForEvent("framereceived"),
+    page.click("button[data-testid=createContainer]"),
+  ]);
+
+  // Waiting on the framereceived event ensures that the websocket receives events
+  // when resources are created.
+
+  // Make sure the container can be removed.
+  await page.waitForSelector("button[data-testid=deleteContainer]");
+
+  // Clean up
+  await Promise.all([
+    page.click("button[data-testid=deleteContainer]"),
+    page.click("button[data-testid=disconnectSocket]"),
+  ]);
```

Comment:
Then you could also assert down here that you've actually received the deleteContainer event, before you disconnect the socket.

Reply:
c6aa1edf670687ffab221f6e2e2b2ac6446874f9